### DISCUSSION TITLE
HOWTO import bookmarks from Wallabag

### DISCUSSION THREAD

**Discussion by rkraken** [karakeep-app/karakeep#581] [General]
After self hosting Wallabag for years, I'm switching to Hoarder.  The modern UI and AI tagging features are great.   In case it helps anyone, this is what I did to bring over my bookmarks from Wallabag.

First, create any desired custom AI tagging rules in Hoarder.

In Wallabag, export bookmarks as .json format.

Using json to html script from https://github.com/KillianKemps/wallabag-to-html, first install Ruby:
```
sudo apt-get update && sudo apt-get install ruby-full

gem install sanitize
```
Download the script json_to_html_bookmarks.rb from the site.  Then
```
ruby json_to_html_bookmarks.rb your_wallabag_export.json > exported_bookmarks.html
```

Theoretically you should be done here and be able to import this html file into Hoarder.  But if you get an error like I did, continue on.  

Extract links from the HTML file (from Hoarder docs):
```
cat exported_bookmarks.html | grep HREF | sed 's/.*HREF="\([^"]*\)".*/\1/' > all_links.txt
```

Create an api key from your hoarder user settings for this last step.

Import the links file to Hoarder.  This command combines the docker based hoarder CLI and import command from Hoarder Docs:
```
cat all_links.txt | xargs -I{} docker run --rm --network host ghcr.io/hoarder-app/hoarder-cli:release --api-key <key> --server-addr http://localhost:3000 bookmarks add --link {}
```
You will have to change the port number in the command if it was remapped.

All done!

**Comment by MohamedBassem** [comment]
@rkraken thanks for the writeup!

I had a look at https://github.com/KillianKemps/wallabag-to-html and it seems like it converts it to the netscape format which hoarder suppots natively. So once you export from wallabag and run the ruby script to convert it, I'd recommend importing it from the user settings page instead. The benefit of this over using the CLI is that if you import it from the web app, hoarder will carry forward your tags, titles, and created At dates.

**Reply by rkraken** [reply]
Hi @MohamedBassem - I actually did try to import the html file but got an error.  If I remember it was something about not being bookmarks format.  Just updated my post.

It would have been nice to keep all my tags, but on the other hand, I guess it will force me to do look through all my stuff and do some well overdue house cleaning.

**Reply by axilleas** [reply]
I used the same script to import my wallabag bookmarks and got the same error. I tried to import the exported bookmarks from Firefox and it worked, so then I started comparing the HTML files.

After digging into the hoarder source code, I saw this:

https://github.com/hoarder-app/hoarder/blob/0c4ea74fb7cff67631152002915df12653896708/apps/web/lib/importBookmarkParser.ts#L25

It seems the Ruby script creates a blank line at the top of the export file, and hoarder doesn't like this. After deleting the blank line, the bookmarks got imported like @MohamedBassem suggested :smiley: 

@rkraken you also have a typo at the convert command, there shouldn't be a `>`.  The usage is:

```shell
ruby json_to_html_bookmarks.rb wallabag.json bookmarks.html
```

And since we don't need the item description of the wallabag entries, the Ruby script could be simplified to:

```ruby
#!/usr/bin/env ruby
require 'json'
require 'time'

abort "Usage: ./script import-file [export-file]" if ARGV.empty?

input_file = ARGV.shift
export_filename = ARGV.shift
export_filename = 'exported_bookmarks.html' if export_filename.empty? || export_filename.nil?

loaded_file = JSON.load(File.open(input_file))

output_begin = '''<!DOCTYPE NETSCAPE-Bookmark-file-1>
<META HTTP-EQUIV="Content-Type" CONTENT="text/html; charset=UTF-8">
<TITLE>Bookmarks</TITLE>
<H1>Bookmarks</H1>
<DL><p>
'''

output_end = '</DL><p>'

output = output_begin

for item in loaded_file do
  article = "<DT><A HREF=\"#{item['url']}\" ADD_DATE=\"#{Time.parse(item['created_at']).to_i}\" LAST_MODIFIED=\"#{Time.parse(item['updated_at']).to_i}\" TAGS=\"#{item['tags'].join(',')}\">#{item['title']}</A>\n"
  output << article
end

output << output_end

output_file = File.open(export_filename, 'w')

output_file.write(output)

output_file.close
```

In which case, you don't even need to install `sanitize`, and the blank line is fixed.

**Reply by MartenBE** [reply]
@axilleas This works great!

**Comment by Stehos** [comment]
You can also download json from your wallabag account and then open chatgpt and promt to convert your json into html 

`can you convert this to html format so I can import my bookmarks into karakeep account?`
